EBIvariation/eva-accession
# Mongodb item writer implementation

### Revision 4

- Add missing processor tests
- Add RS ID list to logger message for documents that didn't persist in the collection, and raised `BulkOperationException`
- Address PR comments (undo copyright changes for non-new files, 25L, unused imports, class doc for writer)

#### Test
- Unit test null-case for processor
- Unit test attributes verification for processed JSON after processor sub-step
- Sample RSID message looks like `RS IDs of duplicate documents: [3906, 3908, 34532]` (might need rewording)

### Revision 3

- Add writer based on `DbsnpClusteredVariantWriter`
- Add steps test (job execution flow)
- Add writer test

#### Test

- Run the job. Result: 25 variants insert.
- Delete 5 variants. code: `db.dbsnpClusteredVariantEntity.remove( {type : {$eq: "MNV"}})`. Result: 20 variants.
- Re-run the job. Result: 25 variants.
- Check database. Missing data added.

### Revision 2

- Refactor `DbsnpClusteredVariantWriter to `eva-accession-core`
- Refactor `ImportCounts` to `eva-accession-core` (why this approach? because otherwise `import-dbsnp2` pipeline would need to depend on `ImportCounts` of eva-accession-import` or have a copy of it
- Add step progress listener

#### Notes

I tried the following code, and it works for my pipeline but breaks existing test cases:
So for now, have not added this (deleting/modifying existing test cases seems risky so)

```
try {
   importedClusteredVariants.stream().filter(entry -> entry != null).forEach(entry -> {
       DBObject dbDoc = new BasicDBObject();
       mongoTemplate.getConverter().write(entry, dbDoc);
       mongoTemplate.save(dbDoc, "dbsnpClusteredVariantEntity");
   });
   importCounts.addClusteredVariantsWritten(importedClusteredVariants.size());
} catch (BulkOperationException e) {
   BulkWriteResult bulkWriteResult = e.getResult();
   importCounts.addClusteredVariantsWritten(bulkWriteResult.getInsertedCount());
   throw e;
}
``` 
Caveat: If values exist already, Job exits with `BulkWriteError ... E11000 duplicate key error ...`. To pass travis for now have disabled 3 test cases in `import-dbsnp2` pipeline (Job force restart true, false, and empty respectively) 

### Revision 1
- Writer implementation

#### Test
- Collection updates with 25 documents in it. Rerun of job results in same 25 documents in collection.
Sample document:
```
{
	 "_id" : "13CB23F25B5A21C0D13B839F43A31CD50BD49414", 
	"_class" : "uk.ac.ebi.eva.accession.core.persistence.DbsnpClusteredVariantEntity", 
	"asm" : "GCF_000001405.38", 
	"tax" : 9606, 
	"contig" : "NC_000024.10", 
	"start" : NumberLong(5793498), 
	"type" : "MNV", 
	"accession" : NumberLong(34072347), 
	"version" : 1, 
	"createdDate" : ISODate("2006-03-13T11:23:00Z") 
}
```
